Multiple Custom Themes (searching the themes folder for directories with theme.json and lone qss files)

TayouVR opened the pull request:
added theme scanning support
tested with various themes from obs and modorganizer2

The themes(qss file and folder with svg graphics for controls) would be put in Prism Launcher's data folder /themes

![image](https://user-images.githubusercontent.com/31988415/196820958-ce9fb1c2-db51-41d8-ada5-dc11b96ef1c3.png)

I just kind of fiddled this into the existing code for the Custom Theme, so if I should clean something still up or load files in a different structure (e.g. one folder per theme) I will do that!

forkiesassds commented on launcher/ui/themes/CustomTheme.cpp:
```diff
@@ -3,192 +3,36 @@
+    if(!FS::ensureFilePathExists(path))
     {
-        qWarning() << "couldn't create folder for theme!";
+        qWarning() << m_name + " Theme file pathdoesn't exist!";
```
There seems to be a missing space

TayouVR replied:
will fix, thank you!

Derpitron commented:
Perhaps this could lead to making a community repo for Prism Launcher themes? I have a Nord Theme which I would love to share with others!

TayouVR commented:
I have thought about how the themes folder could work and I am going to make it look for qss files recursively. That way one could put all the themes from OBS into a OBS folder and other themes with the same name could hopefully exist as well (probably need to compare new names with existing ones and append something to show they are different.

I will try to implement that today hopefully

32bitx64bit reviewed (approved):
:+1:

Scrumplex commented:
I would propose an alternative. We should keep the index.json, but maybe add a `formatVersion` so that we can extend the format. Then we can have `simple` themes which consist of the current format, and we could have `qss` themes that are more advanced.

TayouVR commented:
I was more thinking of doing what was in the json as a separate thing aka .colors files..
I'll think about this for a bit...

for reference, [these](https://github.com/vinceliuice/Fluent-kde/tree/main/color-schemes) kind of files are included in many themes as color palettes, so I was thinking below icon and widget themes we could have another selector for color themes/palettes. This would mean building a system to read those ini style files and loading them.
I could build something to convert the json file colors into that .colors format.

But I'll think about it more..

AshtakaOOf commented:
Would it be possible to make this PR add more icons packs to the icons setting, instead of only having one icon pack at a time?
Instead of adding only theme scanning, it would be nice if you did that?

TayouVR commented:
that would be possible, though its a different chunk of code I would have to disect. I will have a look through it.

flowln reviewed (commented):
Currently, this kinda breaks existing themes. You mentioned doing something with .colors files, so I suppose it has something to do with that (the `theme.json` is not being used here, so custom colors / theme name / theme parent aren't loaded). Are you still going to do that? If so, maybe you should convert this to a draft :>

Also, things with icon packs can be done in a separate PR if you really want to work on those :p

Derpitron commented:
> Currently, this kinda breaks existing themes.

The current theme "system" is very barebones, fit only for minor css color edits. Especially now that theming itself is only in infancy, I think it's the best time to make major overhauls to the theming system, before more mature themes are made and it's too late to change. As a theme creator, I'm perfectly willing to rework my theme to fit the new format, though I'm not sure about other creators.

TayouVR commented:
> (the `theme.json` is not being used here, so custom colors / theme name / theme parent aren't loaded). Are you still going to do that?

I did re-implement that .json file just now, but the commit isn't final. As it stands all themes should load and all should work well.

Implementing Colors via .colors files is probably a bit out of scope for this PR and changing how icon themes work too probably. So I think I'll leave that up to other people / another PR of my own.
And about the .colors files.. I'm not even sure how common they are, I just saw them being used in a few other themes...

TayouVR commented:
I think I am now satisfied with this PR. 
It will first search for any Directories with a theme.json file, and load the theme for it, then it will scan all remaining directories in the theme folder and load qss and css files directly as themes

TayouVR committed:
add support for multiple custom themes

also moved theme related code from Application.cpp to new ui/themes/ThemeManager.cpp, this class should cleanly isolate theme related functions and help avoid code duplication in future theme related additions.
Themes can now be just qss or css files, they won't have color pallette information with them in that case

Signed-off-by: Tayou <tayou@gmx.net>

FederAndInk commented on launcher/Application.h:
```diff
@@ -257,9 +262,9 @@ private slots:
-    std::map<QString, std::unique_ptr<ITheme>> m_themes;
     std::unique_ptr<MCEditTool> m_mcedit;
     QSet<QString> m_features;
+    ThemeManager* m_themeManager;
```
why not use a `std::unique_ptr`?

FederAndInk commented on launcher/Application.cpp:
```diff
@@ -748,28 +745,11 @@ Application::Application(int &argc, char **argv) : QApplication(argc, argv)
-        QIcon::setThemeSearchPaths(searchPaths);
-        qDebug() << "<> Icon themes initialized.";
-    }
+        m_themeManager = new ThemeManager(m_mainWindow);
```
I would use `std::make_unique` + a `std::unique_ptr`

Scrumplex commented on launcher/Application.cpp:
```diff
@@ -1,7 +1,10 @@
 // SPDX-License-Identifier: GPL-3.0-only
 /*
- *  PolyMC - Minecraft Launcher
+ *  Prism Launcher
  *  Copyright (C) 2022 Sefa Eyeoglu <contact@scrumplex.net>
+ *  Copyright (C) 2022 Tayou <tayou@gmx.net>
+ *
+ *  PolyMC - Minecraft Launcher
  *  Copyright (C) 2022 Lenny McLennington <lenny@sneed.church>
```
Just change `PolyMC` to `Prism Launcher` and keep the existing list of copyright holders

```suggestion
 *  Prism Launcher - Minecraft Launcher
 *  Copyright (C) 2022 Sefa Eyeoglu <contact@scrumplex.net>
 *  Copyright (C) 2022 Lenny McLennington <lenny@sneed.church>
 *  Copyright (C) 2022 Tayou <tayou@gmx.net>
```

Scrumplex replied:
Accidental approval lol

TayouVR replied:
already changed locally. 
~~the QList stuff we talked about last afternoon broke some things, which is why i didn't commit & push yet.
I think I'm just gonna not do that for now, its staying a std::vector ^^~~
Was crashing because of something else, maybe I'll do the QList stuff again then

FederAndInk commented on launcher/ui/themes/CustomTheme.h:
```diff
@@ -1,11 +1,51 @@
 
 #include "ITheme.h"
+#include <QFile>
+#include <QDir>
```
might not be needed in the header, I would forward declare `QFileInfo` and include that in the source file. but this is a minor thing

FederAndInk commented on launcher/ui/themes/ThemeManager.cpp:
```diff
@@ -0,0 +1,133 @@
+    return m_themes[themeId].get();
+}
+
+void ThemeManager::InitializeThemes() {
```
why separate initialization from the ctor? I might not see the entire picture, but I believe it could be in the ctor to avoid forgetting to call this.

FederAndInk commented on launcher/ui/themes/ThemeManager.cpp:
```diff
@@ -0,0 +1,133 @@
+        themeDebugLog "✓ Loading Built-in Theme:" << AddTheme(new BrightTheme());
+
+        // TODO: need some way to differentiate same name themes in different subdirectories (maybe smaller grey text next to theme name in dropdown? Dunno how to do that though)
+        QString themeFolder = (new QDir("./themes/"))->absoluteFilePath("");
```
It seems like you are leaking memory there, you should be able to simply do `QString themeFolder = QDir("./themes/").absoluteFilePath("");`

FederAndInk commented on launcher/ui/themes/ThemeManager.cpp:
```diff
@@ -0,0 +1,133 @@
+std::vector<ITheme *> ThemeManager::getValidApplicationThemes()
+{
+    std::vector<ITheme *> ret;
+    auto iter = m_themes.cbegin();
+    while (iter != m_themes.cend())
+    {
+        ret.push_back((*iter).second.get());
+        iter++;
+    }
```
You should be able to use a range for loop: https://en.cppreference.com/w/cpp/language/range-for or even use a provided algorithm ^^ https://en.cppreference.com/w/cpp/algorithm with `std::back_inserter`

TayouVR replied:
this part was code I had moved from Application.cpp, so I haven't actually looked into optimizing it, I will have a look at the two links you sent though and see what I can do ^^

TayouVR replied:
I've changed the loop to be a "for each", this seems to compile and run fine, but I don't actually know what the functions in the loop did or do now. 
Like.. I know it puts the Themes into a "List" (Vector), but I don't actually know what e.g. emplace does xD
```cpp
std::vector<ITheme *> ThemeManager::getValidApplicationThemes()
{
    std::vector<ITheme *> ret;
    ret.reserve(m_themes.size());
    for (auto&& [id, theme] : m_themes) {
        ret.emplace_back(theme);
    }
    return ret;
}
```

FederAndInk replied:
this is called a range-for loop :)
emplace construct the element in place in the vector instead of making a copy. it forwards the argument until it can construct the element in the vector directly.

btw, it looks strange that it compiles as is, as theme should be a `std::unique_ptr`, so you would need to do `.get()` on it to have a raw pointer

FederAndInk replied:
in this case, emplace_back is not really doing anything different from push_back, as you are putting raw pointers in the vector

TayouVR replied:
I will add the .get(), but it seemed to work fine without it too.

FederAndInk commented on launcher/ui/themes/ThemeManager.cpp:
```diff
@@ -0,0 +1,133 @@
+
+std::vector<ITheme *> ThemeManager::getValidApplicationThemes()
+{
+    std::vector<ITheme *> ret;
```
you can reserve the space as you know the output size :)
`ret.reserve(m_themes.size())`

FederAndInk commented on launcher/ui/themes/ThemeManager.cpp:
```diff
@@ -0,0 +1,133 @@
+    auto themeIter = m_themes.find(name);
+    if(themeIter != m_themes.end())
+    {
+        auto & theme = (*themeIter).second;
```
you can use `themeIter->second`

FederAndInk commented on launcher/ui/themes/ThemeManager.h:
```diff
@@ -0,0 +1,47 @@
+#include "ui/themes/ITheme.h"
+#include "ui/MainWindow.h"
+
+#define themeDebugLog qDebug() << "[Themes]"
+#define themeWarningLog qWarning() << "[Themes]"
```
I would use inline functions to avoid defining macros when you have alternatives :)

TayouVR replied:
I thought this would be a neat way of doing it, coming from C# and java land I would have done it via functions, just thought this was cleaner or more how its supposed to be done in C++
I guess i can change it to a function.

FederAndInk replied:
You just need to use an inline function :slightly_smiling_face: 
In this case, the compiler will surely optimize the call.
Macros can have unexpected side effects and can lead to really weird errors, it's basically text replacement. It's more a C thing but in C++ we have been replacing them more and more: templates, constexpr, inline, variadic templates, ... :slightly_smiling_face:

FederAndInk replied:
you can just define it like so:
```cpp
inline auto themeDebugLog() {
  return qDebug() << "[Theme]";
}
```

FederAndInk commented on launcher/ui/themes/ThemeManager.cpp:
```diff
@@ -0,0 +1,133 @@
+/// @brief Adds the Theme to the list of themes
+/// @param theme The Theme to add
+/// @return Theme ID
+QString ThemeManager::AddTheme(ITheme * theme) {
```
I would make this take a `std::unique_ptr<ITheme>` and then `std::move` it into the map to clearly state that this is taking ownership.
Also, I think you can return a `QString const&` instead of a copy or a `QStringView`

FederAndInk commented on launcher/ui/themes/ThemeManager.cpp:
```diff
@@ -0,0 +1,133 @@
+/// @param theme The Theme to add
+/// @return Theme ID
+QString ThemeManager::AddTheme(ITheme * theme) {
+    m_themes.insert(std::make_pair(theme->id(), std::unique_ptr<ITheme>(theme)));
```
I believe C++17 is available, if so can use [emplace()](https://en.cppreference.com/w/cpp/container/map/emplace) to avoid the make_pair, or with insert you can call the ctor of pair implicitly e.g: `insert({first, second})`

It would give something like this:

```cpp
QString const& ThemeManager::AddTheme(std::unique_ptr<ITheme>* theme) {
    QString const& id = theme->id();
    m_themes.emplace(id, std::move(theme));
    return id;
}
```

or replacing `QString const&` by `QStringView`

FederAndInk commented on launcher/Application.cpp:
```diff
@@ -879,6 +859,10 @@ Application::Application(int &argc, char **argv) : QApplication(argc, argv)
     performMainStartupAction();
 }
 
+ThemeManager* Application::getThemeManager() {
```
Could also return a reference, I don't have a strong opinion on this one. depends on the use of the function I guess.

TayouVR replied:
I could also remove the function for now as nobody is asking for it, I just put it as other parts of the code might want to ask for it. (e.g. if I were to remove `getValidApplicationThemes` and `SetApplicationThemes` from Application.cpp)

TayouVR replied:
I removed it. It wasn't useful and didn't play nice with the unique_ptr

FederAndInk replied:
> didn't play nice with the unique_ptr

what do you mean?

you can just do this:
```cpp
ThemeManager& Application::getThemeManager() {
  return *m_themeManager;
}
```

with `m_themeManager` being a `std::unique_ptr<ThemeManager>`

TayouVR replied:
can always do that when its actually needed. As it was it was just a dead function anyways.

FederAndInk reviewed (commented):
Some general comments others can react to ^^
Not really a contributor to prism, but I really like it :heart:
and as a C++ dev, I can give some advice or rise questions, feel free to tell me not to :)
also, I can recommend running clang-format on the changes, I've seen a few inconsistencies in the styling.

32bitx64bit reviewed (changes requested):
Changing all the copyright to prism launcher isnt very legal sense this is a fork of PolyMC

Could present legal issues later on.

Scrumplex commented:
> Changing all the copyright to prism launcher isnt very legal sense this is a fork of PolyMC
> 
> Could present legal issues later on.

No. The project name is not part of the copyright notice. It is just to inform the user what project the current file comes from. As long as the actual copyright holders' notices are preserved, it's fine.

32bitx64bit reviewed (approved):
Looks good now

flowln commented on launcher/ui/themes/CustomTheme.cpp:
```diff
@@ -15,8 +52,11 @@ static bool readThemeJson(const QString &path, QPalette &palette, double &fadeAm
             name = Json::requireString(root, "name", "Theme name");
             widgets = Json::requireString(root, "widgets", "Qt widget theme");
+            qssFilePath = Json::ensureString(root, "qssFilePath", "themeStyle.css");
+            //auto colorFileList = Json::ensureArray(root, "colorFiles");
```
```suggestion
```

dead code must go :gun:

TayouVR replied:
guess I'll save that in stash for the next PR then, fair. x3

flowln commented on launcher/ui/themes/CustomTheme.cpp:
```diff
@@ -119,75 +160,112 @@ static bool writeThemeJson(const QString &path, const QPalette &palette, double
 }
 
-CustomTheme::CustomTheme(ITheme* baseTheme, QString folder)
+/// @brief 
```
very brief indeed :thinking: 
does doxygen (i think this is the comment style for them, right?) not work if there isn't a `@brief`?

TayouVR replied:
VSCode auto-generated it, I can probably remove the brief, I'll have a look

FederAndInk commented on launcher/Application.cpp:
```diff
@@ -746,11 +744,7 @@ Application::Application(int &argc, char **argv) : QApplication(argc, argv)
-
-        m_themeManager->InitializeThemes();
-    }
+    m_themeManager = std::make_unique<ThemeManager>(ThemeManager(m_mainWindow));
```
here you just need to do `m_themeManager = std::make_unique<ThemeManager>(m_mainWindow);`

TayouVR commented on launcher/ui/themes/ThemeManager.cpp:
```diff
@@ -55,55 +57,51 @@ void ThemeManager::InitializeThemes() {
-        themeDebugLog "✓ Loading Built-in Theme:" << darkThemeId;
-        themeDebugLog "✓ Loading Built-in Theme:" << AddTheme(new BrightTheme());
+        themeDebugLog() << "<> Initializing Widget Themes";
+        themeDebugLog() << "Loading Built-in Theme:" << AddTheme(std::make_unique<SystemTheme>(SystemTheme()));
```
this is crashing, I'm not doing something right....

can anyone help me here?
I changed AddTheme() from `ITheme*` to `std::unique_ptr<ITheme>` as you can see above, but doing so means I need to change how I'm adding themes... apparently not like this though as it crashes as soon as it tries to load the first one.

TayouVR replied:
The return type is the problem, I couldn't return a ref without further changes in ITheme, so I'm leaving it as simple QString for now.

FederAndInk commented on launcher/ui/themes/ThemeManager.cpp:
```diff
@@ -55,55 +57,51 @@ void ThemeManager::InitializeThemes() {
-        themeDebugLog "✓ Loading Built-in Theme:" << darkThemeId;
-        themeDebugLog "✓ Loading Built-in Theme:" << AddTheme(new BrightTheme());
+        themeDebugLog() << "<> Initializing Widget Themes";
+        themeDebugLog() << "Loading Built-in Theme:" << AddTheme(std::make_unique<SystemTheme>(SystemTheme()));
```
same here, just `std::make_unique<SystemTheme>()`
and same for the following ones

FederAndInk commented on launcher/ui/themes/ThemeManager.cpp:
```diff
@@ -28,14 +28,16 @@
-    m_themes.insert(std::make_pair(theme->id(), std::unique_ptr<ITheme>(theme)));
-    return theme->id();
+QString const& ThemeManager::AddTheme(std::unique_ptr<ITheme> theme) {
+    QString const& id = theme->id();
+    m_themes.emplace(id, std::move(theme));
+    return id;
```
Oh wait, `theme->id()` returns a copy :/
didn't know that, so you can't return id by ref. either make `id()` return a ref or return by copy, your crash is probably here. sorry for the bad suggestion

TayouVR replied:
all good, its fixed now hopefully at least. ^^

TayouVR committed:
condensed icon theme code

Signed-off-by: Tayou <tayou@gmx.net>

FederAndInk commented on launcher/ui/pages/global/LauncherPage.cpp:
```diff
@@ -383,45 +383,12 @@ void LauncherPage::loadSettings()
+    for (qsizetype i = 0; i < iconThemeOptions.count(); i++) {
+        if (iconThemeOptions[i] == theme) {
+            ui->themeComboBox->setCurrentIndex(i);
+        }
```
`iconThemeOptions.indexOf(theme)` could be used to replace the whole loop and stop before the end

FederAndInk commented on launcher/ui/pages/global/LauncherPage.cpp:
```diff
@@ -383,45 +383,12 @@ void LauncherPage::loadSettings()
-    {
-        ui->themeComboBox->setCurrentIndex(9);
+    QStringList iconThemeOptions;
+    iconThemeOptions << "pe_colored" << "pe_light" << "pe_dark" << "pe_blue" << "OSX" << "iOS" << "flat" << "flat_white" << "multimc" << "custom";
```
The ctor of QList can take a `std::initializer_list` so you can do:

```suggestion
    QStringList iconThemeOptions{"pe_colored", "pe_light", "pe_dark", "pe_blue", "OSX", "iOS", "flat", "flat_white", "multimc", "custom"};
```

FederAndInk commented on launcher/ui/themes/ThemeManager.cpp:
```diff
@@ -0,0 +1,132 @@
+#include "ui/WinDarkmode.h"
+#endif
+
+ThemeManager::ThemeManager(MainWindow* mainWindow) {
+    m_mainWindow = mainWindow;
```
btw, you can use the initializer list :)
for a pointer it won't change a lot though, but still good practice.
```suggestion
ThemeManager::ThemeManager(MainWindow* mainWindow) : m_mainWindow{ mainWindow }
{
```

TayouVR replied:
i did actually see that before, in other places in code and wanted to use it, but couldn't find it anymore when I was working on that part. thank you!

FederAndInk replied:
you can see that here :)
https://en.cppreference.com/w/cpp/language/constructor

FederAndInk replied:
you can see the example with comments: https://en.cppreference.com/w/cpp/language/constructor#Example

TayouVR committed:
added suggested changes

Signed-off-by: Tayou <tayou@gmx.net>

TayouVR commented:
had done a mistake when rebasing before, fixed now and rebased on newest develop

EDIT: look at those failing builds, this is what I get for not building locally before pushing....

FederAndInk commented on launcher/ui/themes/ThemeManager.cpp:
```diff
@@ -0,0 +1,154 @@
+ */
+#include "ThemeManager.h"
+
+#include <versionhelpers.h>
```
You need to remove this include as this is a windows only file and you've already included it after:

https://github.com/PrismLauncher/PrismLauncher/blob/55881674be900625ab57e982a72262c820719349/launcher/ui/themes/ThemeManager.cpp#L32-L35

TayouVR replied:
ahh, I messed it up when rebasing and moving my local branches around... 😭

FederAndInk replied:
no worries, just pointing this out ^^

FederAndInk commented on launcher/ui/themes/ThemeManager.h:
```diff
@@ -0,0 +1,51 @@
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
```
you forgot `#pragma once`

FederAndInk commented on launcher/Application.h:
```diff
@@ -44,6 +45,8 @@
 #include <QUrl>
 #include <updater/GoUpdate.h>
 
+#include "ui/themes/ThemeManager.h"
```
You can just forward declare instead, as you are just declaring the pointer in this header, see:
https://github.com/PrismLauncher/PrismLauncher/blob/3c571c074f0544a674671dd6e7f45ecf84f0c06b/launcher/Application.h#L54

then you can include this in Application.cpp

TayouVR replied:
so I would do this?
```suggestion
class ThemeManager;"
```
in Application.h and move the #include to the .cpp file?

I needed to look up what forward declaring something means and why you could need it.

In this case I suppose it would just mean, that other parts of code, which #include Application.h wouldn't automatically include the entire ThemeManager?

FederAndInk replied:
yeah, you would put `class ThemeManager;` after https://github.com/PrismLauncher/PrismLauncher/blob/3c571c074f0544a674671dd6e7f45ecf84f0c06b/launcher/Application.h#L73-L74

then move the `#include "ui/themes/ThemeManager.h"` to Application.cpp

> In this case I suppose it would just mean, that other parts of code, which #include Application.h wouldn't automatically include the entire ThemeManager?

yup, so when you touch ThemeManager.h it just recompiles a few files instead of a hundred ^^

FederAndInk replied:
basically, when declaring a pointer (`A* a;`) or something that doesn't need to know the size of a class like `std::vector<A> v;` you can forward declare A like so `class A;` you need the definition of A when you use it (member access, dereference, ...). when doing forward declaration you avoid including a whole file in a header that would be included in a whole lot of other files, this will avoid recompiling the header of A a lot speeding up recompilation when developing :) but also compilation as you don't compile the header of A too many times.

FederAndInk commented on launcher/ui/themes/ThemeManager.cpp:
```diff
@@ -0,0 +1,153 @@
+#include "Application.h"
+
+#ifdef Q_OS_WIN
+#include <versionhelpers.h>
```
woah, this one is cursed, you actually need to include windows.h before it:

```suggestion
#include <windows.h>

#include <versionhelpers.h>
```

it worked before because windows.h is included in ui/WinDarkmode.h

TayouVR replied:
I was wondering why that was failing, going by the log output i'm guessing the system provided versionhelpers.h is lacking a #include somewhere? because thats where the build errors are coming from..
(that was the one conflict when rebasing, the versionhelpers.h was included in Application.cpp, but apparently doesn't just work on its own... needs windows.h too...)

TayouVR replied:
oh, I see what you mean with the last line of your comment, so changing the include order would fix it, but its cursed and should be included properly ....

FederAndInk replied:
yeah, i'd rather put the windows.h before instead of putting WinDarkmode which include windows.h before ^^
in any case this is cursed, but don't worry, windows.h will be put only once here thanks to header guard/pragma once

TayouVR commented:
I don't like these messy commits and squashed them locally, I'm just making sure i didn't break anything rn via a local build, then push and hopefully finally just merge, I wanna go to the next PR (which *kinda* depends on this) xD

flowln commented on launcher/ui/themes/CustomTheme.cpp:
```diff
@@ -119,75 +160,112 @@ static bool writeThemeJson(const QString &path, const QPalette &palette, double
 
-        QFileInfo info(themeFilePath);
-        if(!info.exists())
+            QFileInfo info(themeFilePath);
```
not used?

TayouVR replied:
might've overlooked

flowln commented on launcher/ui/themes/ThemeManager.cpp:
```diff
@@ -0,0 +1,154 @@
+/// @brief Adds the Theme to the list of themes
+/// @param theme The Theme to add
+/// @return Theme ID
+QString ThemeManager::AddTheme(std::unique_ptr<ITheme> theme) {
```
```suggestion
QString ThemeManager::AddTheme(std::unique_ptr<ITheme>&& theme) {

```

not sure if i'm getting this right. Since you're doing `std::move` from this function, you won't be able to rely on it being valid after the fact, right? So, we can communicate that using a && thingy.

FederAndInk replied:
this is the point of unique_ptr :)
you can only move it, so by copy is the right thing to do to mean this function takes ownership

flowln replied:
but if you only do a copy, you're not signaling to the user of this function that the pointer they passed to it will no longer have the valid object in it :thinking:

FederAndInk replied:
exactly, otherwise you would use a shared_ptr or return a reference. this is just to mean that this is taking ownership. In the usage for now, the pointer is not used after being passed to AddTheme

FederAndInk replied:
in any case with a unique_ptr you should have only 1 owning pointer you can have raw pointers as observers but you shouldn't keep them as the unique_ptr could delete the memory. so it depends on what you want to do, but in this case I find a unique_ptr with the AddTheme working great. as shared_ptr should be aboided if possible

flowln replied:
so, this function is taking ownership of the unique_ptr with or without copying. This means that, regardless of which one we use, it's gonna do the same stuff. The thing is, using && makes the unique_ptr usage similar to any other object you pass to a function that takes ownership of it. Adding the && just makes it more clear that this is going on. Is this right?

FederAndInk replied:
nope ^^
you can't copy a unique_ptr you can only move it
&& is a rvalue reference, it means it can only bind to rvalues, but it's just a reference so if you do that you can move inside the function or you may not, the unique_ptr is already stating the fact that it can only be moved, and by value insist on the fact that it will be moved

FederAndInk replied:
also && is adding a layer of pointer, while just passing by value just copy the pointer

flowln replied:
this screams bad language design, would be much easier to read code if you just had everything that gets moved using the same syntax, even if it technically makes sense to be this way. But I guess it is what it is :woman_shrugging:

FederAndInk replied:
well move semantic wasn't built-in from the start, it was added in C++11 ^^
I guess its a habit, `std::move` doesn't move it just casts to an rvalue reference, && either, it's just the rvalue reference and it binds to rvalues, the actual move occurs when calling the move constructor or move assignment operator, which are called when passing an rvalue of the same type to them. also move is a big word, often its just "stealing" so copy of pointers and zeroing the source.
you can have a good grasp at this with CppCon videos, they are really great, they have a "back to basics" series: https://www.youtube.com/watch?v=knEaMpytRMA&t=6 :)

flowln replied:
CppCon is great, i'll have a look, thanks `\ ^-^/`

flowln reviewed (commented):
i guess there's no way to dynamically remove themes rn in the application, but the fact that the baseTheme on a theme is a raw pointer to a theme that's managed as a unique_ptr in m_themes can come bite us later if we do add this capability :grimacing:

also, why do the themes get based off the dark theme? shouldn't it depend on the theme itself? :thinking: 

with that said, it seems to work pretty fine for me :slightly_smiling_face:

DioEgizio commented on launcher/ui/themes/SystemTheme.cpp:
```diff
@@ -1,30 +1,69 @@
+ *
+ *      Copyright 2013-2021 MultiMC Contributors
+ *
+ *      Authors: Andrew Okin
```
is this just copy pasted? can only find peterix in git history

TayouVR replied:
oh yep, I hadn't noticed the MultiMC part had authors at all listed, I'll fix that!

DioEgizio commented on launcher/ui/themes/CustomTheme.h:
```diff
@@ -1,11 +1,50 @@
+ *
+ *      Copyright 2013-2021 MultiMC Contributors
+ *
+ *      Authors: Andrew Okin
```
same here

DioEgizio commented on launcher/ui/themes/CustomTheme.cpp:
```diff
@@ -1,12 +1,49 @@
+ *
+ *      Copyright 2013-2021 MultiMC Contributors
+ *
+ *      Authors: Andrew Okin
```
same here

Scrumplex commented on launcher/Application.cpp:
```diff
@@ -1,8 +1,9 @@
 // SPDX-License-Identifier: GPL-3.0-only
 /*
- *  PolyMC - Minecraft Launcher
+ *  Prism Launcher
```
```suggestion
 *  Prism Launcher - Minecraft Launcher
```

Scrumplex commented on launcher/Application.h:
```diff
@@ -1,7 +1,8 @@
 // SPDX-License-Identifier: GPL-3.0-only
 /*
- *  PolyMC - Minecraft Launcher
+ *  Prism Launcher
```
yes

Scrumplex commented on launcher/ui/themes/CustomTheme.cpp:
```diff
@@ -1,12 +1,49 @@
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  Prism Launcher
```
yes

Scrumplex commented on launcher/ui/themes/CustomTheme.h:
```diff
@@ -1,11 +1,50 @@
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  Prism Launcher
```
yes

Scrumplex commented on launcher/ui/themes/SystemTheme.cpp:
```diff
@@ -1,30 +1,69 @@
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  Prism Launcher
```
yes

Scrumplex commented on launcher/ui/themes/ThemeManager.cpp:
```diff
@@ -0,0 +1,154 @@
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  Prism Launcher
```
yes

Scrumplex commented on launcher/ui/themes/ThemeManager.h:
```diff
@@ -0,0 +1,51 @@
+// SPDX-License-Identifier: GPL-3.0-only
+/*
+ *  Prism Launcher
```
yes

Scrumplex reviewed (commented):
we need more code review conversations :trollface:

TayouVR commented:
well, looks like I have enough stuff to work on when I get up tomorrow..... 😅

TayouVR committed:
added requested changes

Signed-off-by: Tayou <tayou@gmx.net>

TayouVR committed:
clang_format for PR files

I'm getting a rendezvous... I thought I did this already

Signed-off-by: Tayou <tayou@gmx.net>

TayouVR commented:
I lied, I did it now, at 5:10am....

flowln reviewed (approved):
~101~ 102 comments o.O

this looks pretty good to me, thanks ^^

flowln commented on launcher/Application.cpp:
```diff
@@ -1,8 +1,9 @@
 // SPDX-License-Identifier: GPL-3.0-only
```
temporary conversation because of the accidental Scrumplex approval :p

_hmm yes very interesting SPDX thing_ :monocle_face:

Scrumplex replied:
lol

TayouVR replied:
yes, more conversations please.
![image](https://user-images.githubusercontent.com/31988415/200086425-9e51e2aa-3bde-489b-a859-a425b5e66e8e.png)
at this point I'm becoming a collector lmao

Scrumplex reviewed (approved):
LGTM
